### PR TITLE
Add enhanced template functions using the sprig module

### PR DESCRIPTION
**Pull request by brian-jarvis**
Added additional templating functions by exposing limited functionality from the http://masterminds.github.io/sprig/ library.  This will allow advanced use cases such as getting the cluster name even for local-cluster (ACM-1402), comparing cluster versions (ACM-1402) using semversioning, and decoding json documents stored in configmaps into structures.

Signed-off-by: Brian Jarvis <bjarvis@redhat.com>

### CODE REVIEW

**Comment by mprahl** [comment]
Thanks for the PR @brian-jarvis! I'm a little busy this week so I hope to review this next week.

**Review by dhaiducek — DISMISSED** [review]
/hold for other reviews

Thanks, @brian-jarvis! Sorry for the delay--I think these functions to process input would really enhance the templates!

File: pkg/templates/sprig_wrapper.go [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+// Copyright Contributors to the Open Cluster Management project
+
+package templates
+
+import (
+	"github.com/Masterminds/sprig/v3"
+)
+
+var (
+	SprigFuncMap = sprig.FuncMap()
```
Comment (mprahl) [inline pkg/templates/sprig_wrapper.go:11]: Could you please make this a private variable?
```suggestion
	sprigFuncMap = sprig.FuncMap()
```

File: pkg/templates/sprig_wrapper.go [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+
+	// ExportedSprigFunctions lists all of the functions from sprig that will be exposed
+	ExportedSprigFunctions = []string{
```
Comment (mprahl) [inline pkg/templates/sprig_wrapper.go:14]: Could you please make this a private variable?
```suggestion
	exportedSprigFunctions = []string{
```

Comment (mprahl) [inline pkg/templates/sprig_wrapper.go:14]: Optional: It'd be nice to have this alphabetize so it's easier to check if a function is exported

File: pkg/templates/sprig_wrapper.go [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+		"empty",
+		"fromJson",
+		"mustFromJson",
+		"ternary",
+		"semver",
+		"semverCompare",
+	}
+)
+
+func GetSprigFunc(funcName string) (result map[int]interface{}) {
```
Comment (mprahl) [inline pkg/templates/sprig_wrapper.go:40]: Could you make this a private function?
```suggestion
func getSprigFunc(funcName string) (result map[int]interface{}) {
```

File: pkg/templates/templates.go [hunk 4]
```diff
@@ -346,6 +346,11 @@ func (t *TemplateResolver) ResolveTemplate(tmplJSON []byte, context interface{})
 		"toBool":           toBool,
 	}
 
+	// Add all the functions from sprig we will support
+	for _, fname := range ExportedSprigFunctions {
+		funcMap[fname] = GetSprigFunc(fname)[0]
```
Comment (mprahl) [inline pkg/templates/templates.go:351]: If you're always grabbing the 0th index, could `GetSprigFunc` just return the function directly?

**Comment by brian-jarvis** [comment]
@mprahl  thank you for the input.  All change suggestions should be present.

**Comment by mprahl** [comment]
This looks good @brian-jarvis! Could you please squash your commits? Then it'll be good to merge.

**Comment by brian-jarvis** [comment]
@mprahl  squash pushed.  Thank you

**Review by mprahl — APPROVED** [review]
Nice!

**Pull request merged**